### PR TITLE
[Phone Number Privacy] Restructure query quota and improve logging

### PR DESCRIPTION
**Pull request by jmrossy**
## Description

- Restructure query quota to simplify it
- Improve query quota tests by mocking contract kit instead of local functions
- Add a logging wrapper (logger) that includes a prefix id for each request
- Add more logging throughout

## Tested

Locally and in staging

## Backwards compatibility

Yes

### CODE REVIEW

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts [hunk 1]
```diff
@@ -1,62 +1,62 @@
...
-})
-
-describe('get remaining query count', () => {
-  beforeEach(() => {
-    jest.clearAllMocks()
-  })
-  it('calculates remaining query count', async () => {
+describe(getRemainingQueryCount, () => {
+  it('Calculates remaining query count for verified account', async () => {
+    const contractKitVerifiedNoTx = createMockContractKit(
```
Comment (jmrossy) [inline packages/phone-number-privacy/test/salt-generation/query-quota.test.ts:19]: By mocking contractKit instead of spying on and mocking the internal methods of the former Class, we can exercise the code flows more thoroughly and accurately.

File: packages/phone-number-privacy/src/common/logger.ts [hunk 2]
```diff
@@ -0,0 +1,14 @@
+// tslint:disable: no-console
+import { v4 as uuidv4 } from 'uuid'
+
+// This allows us to differentiate parallel instantiations of this function
+const prefix = uuidv4().slice(0, 8)
```
Comment (aslawson) [inline packages/phone-number-privacy/src/common/logger.ts:5]: oh interesting

File: packages/phone-number-privacy/src/database/wrappers/account.ts [hunk 3]
```diff
@@ -10,42 +11,47 @@ function accounts() {
...
+    logger.error(ErrorMessages.DATABASE_GET_FAILURE, e)
     return 0
   }
 }
 
+async function getAccountExists(account: string): Promise<boolean> {
+  const existingAccountRecord = await accounts()
+    .where(ACCOUNTS_COLUMNS.address, account)
+    .first()
+  return !!existingAccountRecord
```
Comment (aslawson) [inline packages/phone-number-privacy/src/database/wrappers/account.ts:31]: so this double negation is a pattern intended to return true if the record exists, correct?

File: packages/phone-number-privacy/src/database/wrappers/account.ts [hunk 4]
```diff
@@ -10,42 +11,47 @@ function accounts() {
...
-      .where(ACCOUNTS_COLUMNS.address, account)
-      .increment(ACCOUNTS_COLUMNS.numLookups, 1)
+    if (await getAccountExists(account)) {
+      return accounts()
+        .where(ACCOUNTS_COLUMNS.address, account)
+        .increment(ACCOUNTS_COLUMNS.numLookups, 1)
+    } else {
+      const newAccount = new Account(account)
+      newAccount[ACCOUNTS_COLUMNS.numLookups] = 1
+      return updateRecord(newAccount)
```
Comment (aslawson) [inline packages/phone-number-privacy/src/database/wrappers/account.ts]: seems like this method should be renamed `insertRecord`

probably my bad here, but lets update

File: packages/phone-number-privacy/test/salt-generation/query-quota.test.ts (hunk 1 above)
Comment (aslawson) [reply packages/phone-number-privacy/test/salt-generation/query-quota.test.ts:19]: wow -- i'm happy to see you figured how to successfully mock the contractkit.

**Review by aslawson — APPROVED** [review]
lgtm

special thanks for cleaning up the query quota testing 🙏

File: packages/phone-number-privacy/src/database/wrappers/account.ts (hunk 3 above)
Comment (jmrossy) [reply packages/phone-number-privacy/src/database/wrappers/account.ts:31]: Yeah, it's a JS way of coalescing a type to boolean

File: packages/phone-number-privacy/src/database/wrappers/account.ts (hunk 4 above)
Comment (jmrossy) [reply packages/phone-number-privacy/src/database/wrappers/account.ts]: Will do

File: packages/phone-number-privacy/src/common/logger.ts [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+
+const logger = {
```
Comment (nambrot) [inline packages/phone-number-privacy/src/common/logger.ts:7]: Sorry for the drive-by, but have you considered using a library like bunyan? I personally find structured logs very helpful, as it allows operators to extract relevant information without much gymanistics. Here is an example for the attestation-service https://github.com/celo-org/celo-monorepo/blob/master/packages/attestation-service/src/logger.ts#L1

File: packages/phone-number-privacy/src/database/wrappers/account.ts [hunk 6]
```diff
@@ -10,42 +11,47 @@ function accounts() {
...
-}
-
-async function incrementExistingQueryCount(account: string) {
+export async function incrementQueryCount(account: string) {
+  logger.debug('Incrementing query count')
   try {
-    return await accounts()
-      .where(ACCOUNTS_COLUMNS.address, account)
-      .increment(ACCOUNTS_COLUMNS.numLookups, 1)
+    if (await getAccountExists(account)) {
```
Comment (nambrot) [inline packages/phone-number-privacy/src/database/wrappers/account.ts:40]: What do you think about using upsert here?

File: packages/phone-number-privacy/src/common/logger.ts (hunk 5 above)
Comment (jmrossy) [reply packages/phone-number-privacy/src/common/logger.ts:7]: Thanks for the suggestions! Have to disagree with you though.
1. I'm not in favor of adding a lib to replace 10 lines of JS that is doing the job, especially in a security sensitive project
2. Why learn a new abstraction layer for logging when console is working!

File: packages/phone-number-privacy/src/database/wrappers/account.ts (hunk 6 above)
Comment (jmrossy) [reply packages/phone-number-privacy/src/database/wrappers/account.ts:40]: Agreed, we'll do that in a follow-up

File: packages/phone-number-privacy/src/common/logger.ts (hunk 5 above)
Comment (nambrot) [reply packages/phone-number-privacy/src/common/logger.ts:7]: 1. very fair
2. Operators of this service might appreciate structured logging/easy integration with logging providers
